BUG: fix PSD normalization for LombScargle.

jakevdp opened the pull request:
The PSD normalization of the Lomb-Scargle periodogram is incorrect in the case of unequal errors. I found this when exploring the distribution of periodogram peaks under the null hypothesis, which should be distributed as exp(-Z).

The following code shows the issue:

```python
import numpy as np
import matplotlib.pyplot as plt
from astropy.stats import LombScargle

def make_data(N=1000, rseed=None):
    rng = np.random.RandomState(rseed)
    t = 100 * np.random.rand(N)
    dy = 0.1 + 0.1 * rng.rand(N)
    y = dy * rng.randn(N)
    return t, y, dy

t, y, dy = make_data()
freq, power = LombScargle(t, y, dy).autopower(normalization='psd',
                                              maximum_frequency=40)

plt.hist(power, 100, histtype='stepfilled', normed=True, alpha=0.5);
z = np.linspace(0, power.max())
plt.plot(z, np.exp(-z), '-k', lw=2);
```

Here is the result in current master:
![wirjjohob52lgaaaabjru5erkjggg](https://cloud.githubusercontent.com/assets/781659/22068589/56d77e8e-dd4a-11e6-92e4-37c83addc5c8.png)

Here is the result using this PR:
![zkpppca 4qx7v7f dy5c3n22wdpamoiksmjszmmxxf53vgkihiguu8jbrgrakncfxejecp0ezeaouixeqkqknqrkqchqhcrcraqdbgrakfcfxejep8ftpdkj1xyiryaaaaasuvork5cyii](https://cloud.githubusercontent.com/assets/781659/22068596/5dde01b2-dd4a-11e6-8d2c-9f2b93012042.png)

The reason I didn't catch this during the initial contribution is that I was using the scipy output and the FFT-based PSD as reference points for the PSD normalization, and neither can handle heteroscedastic errors.

pllim commented:
Thanks for the fix! Need a change log under "bug fix" for 1.3.1 and possibly a test to ensure this bug does not recur.

jakevdp commented:
For completeness, it might be nice to test that the output distribution matches the expectation for each normalization. I'm currently working on another PR that will add some functionality to compute expected distributions & false alarm probabilities for the lomb-scargle output; I'll plan to add such tests there.

mhvk commented:
OK, good to have correct! As @pllim notes, definitely needs a changelog entry.  Also, while it is fine to leave most testing to another PR, it would be great to add at least a simple one here.

jakevdp commented:
>  Also, while it is fine to leave most testing to another PR, it would be great to add at least a simple one here.

OK, will do.

jakevdp commented:
In the last commit I added some functionality to compute expected distributions of the periodogram, and used these to test the normalizations.

pllim commented:
doctest failed due to encoding error in Python 2. FYI.

mhvk commented on astropy/stats/lombscargle/core.py:
```diff
@@ -356,3 +356,97 @@ def model(self, t, frequency):
+        Expressions used here are adapted from Baluev 2008, Table 1.
+        """
+        # TODO: handle units in z?
+        z = np.asarray(z)
```
This will needlessly break for someone putting in `z` that is has a scale but is otherwise dimensionless (say, percent). My sense would be not to worry about this for now, i.e. I'd remove this statement and its comment, as anything with a true dimension will break below anyway and we can see if it makes sense to address that if someone complains.

jakevdp replied:
The best option would be to strip units if the input is dimensionless, and fail otherwise.

jakevdp replied:
But I'm not sure the best way to do that

pllim replied:
Like this?

```python
if isinstance(z, u.Quantity):
    if z.unit == u.dimensionless_unscaled:  # or use "is"?
        z = z.value
    else:
        raise ValueError('blah blah')
```

jakevdp replied:
Yes, I'll give that a shot.

mhvk replied:
I would really just advocate removing this coercion -- the code below will already fail with a `UnitsError` is `z` is not dimensionless, and will give the correct answer if, e.g., a user passes in `z=np.arange(0,100) * u.percent`. In contrast, for that case with the `np.asarray` clause the returned values are wrong and with @pllim's suggestion, one gets an error for no reason.

More generally, I think that unless you *know* that the code will give the wrong answer for a given type of input, you should *not* do any tests on the input (but then, I really like the ducktyping aspect of python and I've learned over the last years that `Quantity` is amply mature enough that, generally, you simply do not have to worry about using it).

jakevdp commented:
Looks like one of the Travis tests is failing due to differences in ``np.histogram`` in numpy 1.7. I'll put a bit more buffer in there...

jakevdp commented:
I think that should fix the numpy 1.7 test failure.

I also opted to move the statistical functionality into ``astropy.stats.lombscargle.statistics``, so as to not modify the public-facing API for this bug fix. Given that I'd soon like to add other statistics & think about the API for accessing them, this seemed like a better route.

jakevdp commented:
Tests passing — I think this is ready for a final review

mhvk commented:
@jakevdp - looking at this again, and trying to think how this would be used, I wonder if we should not have methods on the `LombScargle` class that call these functions.

I think this has multiple benefits: (1) like elsewhere in astropy, all required code can be imported from the top-level of a module, not several levels down, i.e., for the user things stay neatly together; (2) in the `LombScargle` class, you can deal with `psd` normalisation correctly even for the case that there are units, and thus ensure that `Quantity` input is not just an afterthought (which I find problematic currently); (3) the code for the statistics remains uncluttered with input validation -- it can safely assume it will get `ndarray` as input.

What do you think?

jakevdp commented:
@mhvk – I agree that we should have methods on LombScargle to calculate periodogram statistics. But these stats, the cdf and pdf, are not all that useful because they are stats for a single peak. I would be afraid of someone misinterpreting them and treating them as a distribution for the *highest* peak, which they're not.

On another branch I'm working on implementing several versions of the false alarm probability and an API to access them (likely as methods of ``LombScargle``). I decided to submit this bug fix as a separate PR.

jakevdp commented:
It seems to me that a minor release like 1.3.1 is not the right time to add brand-new user-facing API elements.

jakevdp commented:
Also, throughout the Lomb-Scargle code the Quantity is just an afterthought, so this is pretty consistent :smile:

mhvk commented:
You're right, for a bug fix, we should definitely *not* introduce new features. Indeed, I'm quite sure this won't get past the release managers, so fear the `statistics.py` module has to be removed again. Apologies, really @pllim and I should have picked up on that.

mhvk commented:
> Also, throughout the Lomb-Scargle code the Quantity is just an afterthought, so this is pretty consistent

Somewhat regrettably so, in my opinion, since in the core of astropy `Quantity` is either a first-class citizen, or more often, properly required (as in `coordinates`). It certainly has made my code much more reliable -- wouldn't you want to avoid another mission missing Mars because someone messed up their units?

In any case, it being what it is, for the present code, I would still suggest to remove the check on `Quantity` in both `pdf` and `cdf` and simply state that the input should be an `ndarray`.

pllim commented:
Sorry for the confusion, I am not very familiar with `stats` package and was trying to juggle several things at once. My comments were just a general review, not a thorough one. Still, that's no excuse. I apologize!

jakevdp commented:
OK – so should I remove the statistics.py code & its tests, and just leave the distribution of the normalizations untested – i.e. basically revert to my first commit on the branch?

mhvk commented:
Yes, for 1.3.1 lets just do the correction.  I'll forego my initial request for a regression test, since it is clear that that will come with the rest of the code (in 2.0). As said, apologies for causing you yet more work (and thanks again for bringing this nice contribution to astropy!!).

As for `Quantity` -- part of the problem here is that you do need to link to C(ython) code, so you need to convert somewhere -- and I think in the original submission you found the best solution by doing the conversion all on `LombScargle`. So, let's just continue with that!

jakevdp commented:
OK, I'll revert the changes and save the rest for a later PR.

jakevdp commented:
Is it OK to do a force-push here, or would you prefer I open a new PR?

jakevdp commented:
Another option would be to put the statistics functions within the test files for this fix, and move them elsewhere later. What do you think?

mhvk commented:
Well, my ideal solution would be to keep the test files, but simply use the appropriate `pdf` and `cdf` expression (i.e., just calculate `exp(-z)`), and then in the next PR introduce the functions, and add a few lines to the tests which check that the functions produce the same result.

mhvk commented:
p.s. force-push is fine.

jakevdp commented:
> Well, my ideal solution would be to keep the test files, but simply use the appropriate pdf and cdf expression (i.e., just calculate exp(-z))

But I can't just use ``exp(-z)``, because the distribution is different for each normalization. So I either have a conditional within the test that chooses the right expression, or I have a convenience function elsewhere that encapsulates the conditional.

I'll move the functions into the test file.

jakevdp commented:
OK @mhvk – I removed ``statistics.py`` and moved everything into the test script. Chose not to revert because we wanted to keep the tests rather than just use the first commit.

mhvk commented:
@jakevdp - I just made a PR to your branch (https://github.com/jakevdp/astropy/pull/1) that removes the unit checks -- whatever we decide for the final implementation, they don't belong in the bug fix, which should only check that the normalisation is done correctly.

jakevdp commented:
In that case we could also remove the entire ``test_distribution_units`` function as well, because now it doesn't really have a reason for existing.

mhvk commented:
OK, you're right that might as well be gone too.... So, this all looks OK for the bug fix and I'll merge once the tests have passed.

Again, apologies for the tortuous route of this PR...

jakevdp commented:
No worries, thanks for the comments & contribution!

mhvk commented:
All tests passed, except for OSX which was still waiting. I decided to cancel that, and merge.  Thanks again, @jakevdp!

jakevdp commented:
Thanks!
